SingularityService/src/main/java/com/hubspot/singularity/scheduler/SingularityScheduler.java

```diff
@@ -214,66 +217,81 @@ public void checkForDecomissions(SingularitySchedulerStateCache stateCache) {
+      Optional<SingularityPendingDeploy> maybePendingDeploy = deployManager.getPendingDeploy(requestId);
+      List<SingularityTaskId> matchingTaskIds = getMatchingTaskIds(stateCache, maybeRequest.get().getRequest(), deployKey);
 
-      final List<SingularityTaskId> matchingTaskIds = getMatchingTaskIds(stateCache, updatedRequest, pendingRequest);
+      List<SingularityPendingRequest> effectivePendingRequests = new ArrayList<>();
+      pendingRequestsForDeploy.sort(Comparator.comparingLong(SingularityPendingRequest::getTimestamp));
+      for (SingularityPendingRequest pendingRequest : pendingRequestsForDeploy) {
+        final SingularityRequest updatedRequest = updatedRequest(maybePendingDeploy, pendingRequest, maybeRequest.get());
 
-      final SingularityDeployStatistics deployStatistics = getDeployStatistics(pendingRequest.getRequestId(), pendingRequest.getDeployId());
+        if (!shouldScheduleTasks(updatedRequest, pendingRequest, maybePendingDeploy, maybeRequestDeployState)) {
+          LOG.debug("Pending request {} was obsolete (request {})", pendingRequest, SingularityRequestWithState.getRequestState(maybeRequest));
+          obsoleteRequests++;
+          requestManager.deletePendingRequest(pendingRequest);
+          continue;
+        }
 
-      final RequestState requestState = checkCooldown(maybeRequest.get().getState(), updatedRequest, deployStatistics);
+        int missingInstances = getNumMissingInstances(matchingTaskIds, updatedRequest, pendingRequest, maybePendingDeploy);
+        if (missingInstances == 0 && !matchingTaskIds.isEmpty() && updatedRequest.isScheduled() && pendingRequest.getPendingType() == PendingType.NEW_DEPLOY) {
+          LOG.trace("Holding pending request {} because it is scheduled and has an active task", pendingRequest);
+          heldForScheduledActiveTask++;
+          continue;
+        }
 
-      int numScheduledTasks = scheduleTasks(stateCache, updatedRequest, requestState, deployStatistics, pendingRequest, matchingTaskIds, maybePendingDeploy);
+        if (effectivePendingRequests.isEmpty()) {
+          effectivePendingRequests.add(pendingRequest);
+        } else if (pendingRequest.getPendingType() == PendingType.IMMEDIATE) {
+          effectivePendingRequests.add(pendingRequest);
+        } else if (pendingRequest.getPendingType() == PendingType.ONEOFF) {
+          effectivePendingRequests.add(pendingRequest);
+        } else {
+          // Any other subsequent requests are not honored until after the pending queue is cleared.
+        }
 
-      if (numScheduledTasks == 0 && !matchingTaskIds.isEmpty() && updatedRequest.isScheduled() && pendingRequest.getPendingType() == PendingType.NEW_DEPLOY) {
-        LOG.trace("Holding pending request {} because it is scheduled and has an active task", pendingRequest);
-        heldForScheduledActiveTask++;
-        continue;
+        requestManager.deletePendingRequest(pendingRequest);
```

Comment:
for any pending requests that we are scheduling tasks for, the delete should come _after_ the scheduleTasks call. If we are interrupted between here and scheduleTasks we lose a pending request and could get into an inconsistent state